pkg/commands/exec.go

```diff
@@ -0,0 +1,193 @@
+				return nil
+			}
+
+			if err := o.Init(context.Background(), cmd, args); err != nil {
+				return err
+			}
+			if err := o.Complete(); err != nil {
+				return err
+			}
+			if err := o.Run(); err != nil {
+				return err
+			}
+			return nil
+		},
+		Annotations: map[string]string{
+			types.TagCommandType: types.TypeApp,
+		},
+	}
+	cmd.Flags().BoolVarP(&o.Stdin, "stdin", "i", defaultStdin, "Pass stdin to the container")
+	cmd.Flags().BoolVarP(&o.TTY, "tty", "t", defaultTTY, "Stdin is a TTY")
+	cmd.Flags().Duration(podRunningTimeoutFlag, defaultPodExecTimeout,
+		"The length of time (like 5s, 2m, or 3h, higher than zero) to wait until at least one pod is running",
+	)
+	return cmd
+}
+
+func (o *VelaExecOptions) Init(ctx context.Context, c *cobra.Command, argsIn []string) error {
+	o.Context = ctx
+	o.Cmd = c
+	o.Args = argsIn
+
+	env, err := GetEnv(o.Cmd)
+	if err != nil {
+		return err
+	}
+	app, err := application.Load(env.Name, o.Args[0])
+	if err != nil {
+		return err
+	}
+	o.Env = env
+	o.App = app
+
+	cf := genericclioptions.NewConfigFlags(true)
+	cf.Namespace = &o.Env.Namespace
+	o.f = k8scmdutil.NewFactory(k8scmdutil.NewMatchVersionFlags(cf))
+
+	if o.ClientSet == nil {
+		c, err := kubernetes.NewForConfig(o.VelaC.Config)
+		if err != nil {
+			return err
+		}
+		o.ClientSet = c
+	}
+	return nil
+}
+
+func (o *VelaExecOptions) Complete() error {
+	compName, err := o.askComponent()
+	if err != nil {
+		return err
+	}
+	podName, err := o.getPodName(compName)
+	if err != nil {
+		return err
+	}
+	o.kcExecOptions.StreamOptions.Stdin = o.Stdin
+	o.kcExecOptions.StreamOptions.TTY = o.TTY
+
+	args := make([]string, len(o.Args))
+	copy(args, o.Args)
+	// args for kcExecOptions MUST be in such formart:
+	// [podName, COMMAND...]
+	args[0] = podName
+	return o.kcExecOptions.Complete(o.f, o.Cmd, args, 1)
+}
+
+func (o *VelaExecOptions) askComponent() (string, error) {
+	comps := o.App.GetComponents()
+	if len(comps) == 1 {
+		return comps[0], nil
+	}
+	prompt := &survey.Select{
+		Message: "You have multiple Components in your app. Please choose one component for logs: ",
+		Options: comps,
+	}
+	var compName string
+	err := survey.AskOne(prompt, &compName)
+	if err != nil {
+		return "", fmt.Errorf("choosing component name err %v", err)
+	}
+	return compName, nil
+}
+
+func (o *VelaExecOptions) getPodName(compName string) (string, error) {
+	podList, err := o.ClientSet.CoreV1().Pods(o.Env.Namespace).List(o.Context, v1.ListOptions{
+		LabelSelector: labels.Set(map[string]string{
+			//TODO(roywang) except core workloads, not any workloads will pass these label to pod
+			// find a rigorous way to get pod by compname
+			oam.LabelAppName:      o.App.Name,
+			oam.LabelAppComponent: compName,
+		}).String(),
+	})
+	if err != nil {
+		return "", nil
+	}
+	if podList != nil && len(podList.Items) == 0 {
+		return "", fmt.Errorf("cannot get pods")
+	}
+	for _, p := range podList.Items {
```

Comment:
As you choose to return the first Pod whose name matches or doesn't match the component name, Line 181 to 185 seems not necessary.

Reply:
The concern is to return the name-matched pod as a higher priority.